### ISSUE TITLE
Introduce a conceptual help topic about calling external programs (native applications)

### ISSUE THREAD

**Issue by mklement0** [MicrosoftDocs/PowerShell-Docs#5152]
# Summary of the new document or enhancement

Many special considerations apply when you call an external command-line executable (aka native application / utility), which aren't currently covered comprehensively, in one place:

* That the only data type supported is _text_ (`[string]`), both on input and output, and  how _raw byte data_ is fundamentally unsupported - both when collecting the output in PowerShell and when piping between native programs.

* How there are syntax pitfalls due to PowerShell's extended set of metacharacters (compared to other shells) causing potential misinterpretation of arguments, which must be avoided with _quoting_ (e.g., To pass literal `@foo`, which works _unquoted_ in `cmd.exe` and `bash`, you must use `'@foo'` in PowerShell).
   * How `--%` can be used  (primarily on Windows) to selectively deactivate PowerShell's parsing.

* How "native globbing"  is automatically applied to arguments such as `*.txt` on Unix-like platforms; that is, `*.txt` is implicitly replaced with the array of file names / paths matching that wildcard pattern.

* How output data is sent through the pipeline line by line, resulting in an array of strings (lines), if collected in a variable.

* How stderr (standard error) output is _passed through to the host_ rather than going through PowerShell's error stream and can only be captured with a `2>` redirection.

* How redirections (`>`) generally do _not_ pass the native program's output through as-is, but invariably treat it as `[Console]::OutputEncoding` encoded _text_ that on writing to the target file is written with PowerShell's _default encoding_ (BOM-less UTF-8 in PowerShell 6+, UTF-16LE in Windows PowerShell).

* How external-program calls aren't integrated with PowerShell's error handling and require explicit checking of `$?` / `$LASTEXITCODE` to detect failure, except in PowerShell 7, where pipeline chain operators `&&` and `||` can now be used.

Details of requested document:

- Proposed title: about_Native_Program_Calls
- Propose location in the TOC: Among the `about_* topics
- Target audience: end users
- Purpose or scenario: guidance for invoking native command-line programs
- List of related articles to link to: about_Parsing, about_Quoting_Rules, about_Redirection, about_Pipelines, about_pwsh, about_Operators (section "Pipeline chain operators && and ||")

**Comment by Luiz-Monad** [comment]
Another issue when using native utilities is that for the error stream each returned line is embedded into an ErrorObject, which makes things very slow. 
One would perhaps want a "string" stream for that too.

This issue lead me to develop my solution, a specific native processor with custom streams.
https://github.com/Luiz-Monad/PowershellProcess/blob/master/PowerProcess/InvokeProcessFast.cs

**Comment by yecril71pl** [comment]
`C:\'Program Files\Acme Inc.\Your Terrific App.exe'` is a perfect invocation and it does not use the `&` operator.

**Comment by mklement0** [comment]
True: what matters is whether the token _starts with_ a quote, which in turn only works if the _whole_ token is quoted.

It is for this reason that _compound_ string arguments composed of quoted and unquoted tokens are best avoided in PowerShell - the fact that a _quoted_ first token invariably becomes _its own argument_ is counterintuitive and invites confusion:

```powershell
# OK, but only because the first token is unquoted.
C:\'Program Files'\nodejs\node.exe

# NOT OK: becomes string literal *plus separate argument*, because the first token is *quoted*
'C:\Program Files'\nodejs\node.exe

# ALSO NOT OK, despite &, for the same reason.
& 'C:\Program Files'\nodejs\node.exe
```

**Comment by yecril71pl** [comment]
But that makes `&` *recommended*, not *required* as you have suggested.

**Comment by mklement0** [comment]
It is required for (a) command names / paths  that are _quoted as a whole_ and/or (b) contain _variable references or subexpressions_. I've inserted "(as a whole)" in the initial post to clarify.

**Comment by mklement0** [comment]
In short:

* You're free to _always_ use `&`

* You _must_ use `&` in the scenarios described above.

**Comment by surfingoldelephant** [comment]
I agree that this concept (native commands) as a whole needs to be fleshed out quite a bit. I often see users struggle in this area, especially with argument passing and encoding issues.

Here's bits and pieces of the existing native command documentation spread out across many articles:

- [Running native commands](https://learn.microsoft.com/en-us/powershell/scripting/learn/shell/running-commands#running-native-commands)
- [Legacy language spec definition](https://learn.microsoft.com/en-us/powershell/scripting/lang-spec/chapter-01#1-introduction)
- [Native command argument completer](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/register-argumentcompleter#example-3-register-a-custom-native-argument-completer)
- [`$_`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables#section-1) & [Pipeline chain operators](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_pipeline_chain_operators#long-description) with native commands
- [Native command argument parsing](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parsing#passing-arguments-to-native-commands) and [`$PSNativeCommandArgumentPassing`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables#psnativecommandargumentpassing)
- [`Start-Process`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/start-process#notes) (includes a note on when to avoid it in favor of native command invocation)
- [`Switch-Process`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/switch-process)/[`exec`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_built-in_functions#exec) (Unix)
- [Native command globbing on Unix platforms](https://learn.microsoft.com/en-us/powershell/scripting/whats-new/unix-support#general-unix-interoperability-changes)
- [`$ErrorActioNPreference`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables#erroractionpreference) & [`$PSNativeCommandUseErrorActionPreference`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables#psnativecommanduseerroractionpreference)
- [Trace `ParameterBinding` operations for native commands](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/trace-command#example-3-trace-parameterbinding-operations-for-native-commands)
- [`$OutputEncoding`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables#outputencoding)
- [Raw byte preservation](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_pipelines#using-native-commands-in-the-pipeline) (duplicated [here](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_redirection#example-7-redirecting-binary-data-from-a-native-command))

Consolidating this into a single article (with links to existing articles) would go a long way, I think. As would filling in the gaps for the missing information/guidance mentioned in the OP.

These to me are the most pressing areas that need additional content:

- Invocation methods and execution/blocking behavior
    - `Start-Process` vs native command invocation
- Argument passing (pitfalls)
- Output capture/redirection & encoding/decoding
- Error handling

We should also standardize terminology. Right now, it's inconsistent. E.g., these are all used to describe the same concept throughout the documentation:

- Native command/native application/native program/native executable
- Native external command
- External command/external application/external program/external executable
- External executable file
- Operable program (WinPS `CommandNotFoundException`)
- Executable file/executable program (PS `CommandNotFoundException`)

I suggest "native command" going forward. It's already used in the various formal definitions on the topic and is what's used most commonly in existing documentation. It also follows suit with the internal `NativeCommandProcessor` class name (although, that's not a strong argument in itself).

Thoughts?

1. "Run a native command" means to invoke a command that resolves to a file that's part of the OS/host environment file.
    - With or without an invocation operator.
    - As a bareword, quoted string, `ApplicationInfo` instance or variable with a string/`ApplicationInfo` value.
2. Native commands are represented as an `ApplicationInfo` instance by `Get-Command`.
3. All file types excluding PS files (`.ps1`, etc) can be native commands, not just executables (`PATHEXT`).
    - This is a [contentious topic](https://github.com/PowerShell/PowerShell/issues/12625). 
    - Ultimately, I think the most important distinction here is that of native commands vs `Start-Process`, rather than executable vs non-executable file types. E.g., "run `cmd.exe` as a native command", "run `.\foo.txt` natively", etc refer to the same concept.
    - Runtime behavior is somewhat muddied by the fact `.\foo.txt | Write-Output` errors (and is referred to as a "document") whereas `notepad .\foo.txt | Write-Output` works just fine to force synchronous behavior.
5. `Start-Process` provides an _alternative_ method to running a native command with more granular control over execution behavior, but has its own set of drawbacks/pitfalls.
